### PR TITLE
Support SLS logstore

### LINKED ISSUE #67: Which cloud resource would you expect to be supported by provider alibaba?
We are currently implementing Alibaba Cloud OSS (#59 ), SLS project (#58 ) resources, and we are going to support more. 

We are looking forward to your feedback on the cloud resources you want most. 

Cloud resources implementation status:

- [x] ApsaraDB RDS
   - [x] PostgreSQL/Mysql
   - [x] Redis #88  @xiqiu618 @zzxwill 
- [x] SLS
   - [x] SLS Project: #58 
   - [x] SLS Logstore:  #77
   - [x] SLS Logtail 
- [x] OSS Bucket: Done. #59  @zzxwill 
- [ ] MQ: pending.
- [x] NAS
   - [x] FileSystem: #79 @zzxwill 
   - [x] MountTarget: #80 @zzxwill 
- [x] SLB
  - [x] CLB: #84  Update is not yet supported.

### PR DESCRIPTION
**Pull request by zzxwill**
### Description of your changes

Implemented SLS logstore.

Fixes #67

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Apply examples/sls/store.yaml

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: apis/sls/v1alpha1/store_types.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	ShardCount int `json:"shardCount"`
+
+	// Specifies whether to enable automatic sharding. Default value: false.
+	// +optional
+	// +kubebuilder:default:=false
+	AutoSplit bool `json:"autoSplit"`
+
+	// The maximum number of shards for automatic sharding.
+	// +optional
+	MaxSplitShard int `json:"maxSplitShard"`
```
Comment (hasheddan) [inline apis/sls/v1alpha1/store_types.go:73]: By convention, we typically make optional parameters pointers and label with `omitempty`. This helps distinguish whether the field was set to the zero-value of the type (i.e. `false`) or whether it was omitted entirely.
```suggestion
	// Specifies whether to enable automatic sharding. Default value: false.
	// +optional
	// +kubebuilder:default:=false
	AutoSplit *bool `json:"autoSplit,omitempty"`

	// The maximum number of shards for automatic sharding.
	// +optional
	MaxSplitShard *int `json:"maxSplitShard,omitempty"`
```

File: pkg/controller/sls/store_controller.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+	if !ok {
+		return nil, errors.New(errNotStore)
+	}
+
+	var (
+		sel    *xpv1.SecretKeySelector
+		region string
+	)
+
+	switch {
```
Comment (hasheddan) [inline pkg/controller/sls/store_controller.go:81]: It looks like this is being replicated in a number of controllers, might make sense to move it to a common package, such as `client`. `crossplane-runtime` also supports a common credentials extraction mechanism now that will allow you to fetch credentials from a variety of sources. Example here: https://github.com/crossplane/provider-aws/blob/3243f6c4cfa8c860ceee705c9eca72ea68398545/pkg/clients/aws.go#L105

Looks like there might need to be a crossplane-runtime update here first though so might make sense to tackle in a follow-up.

File: pkg/controller/sls/store_controller.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+func (e *storeExternal) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mg.(*slsv1alpha1.Store)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errNotStore)
+	}
+	name := meta.GetExternalName(cr)
+	spec := cr.Spec.ForProvider
+	cr.Status.SetConditions(xpv1.Creating())
+	err := e.client.UpdateStore(spec.ProjectName, name, spec.TTL)
```
Comment (hasheddan) [inline pkg/controller/sls/store_controller.go:173]: nitpick, feel free to disregard if not your preferred pattern :)
```suggestion
	cr.Status.SetConditions(xpv1.Creating())
	err := e.client.UpdateStore(cr.Spec.ForProvider.ProjectName, meta.GetExternalName(cr), cr.Spec.ForProvider.TTL)
```

File: examples/sls/store.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+apiVersion: sls.alibaba.crossplane.io/v1alpha1
+kind: Store
+metadata:
+  name: sls-store-test
+  namespace: default
```
Comment (hasheddan) [inline examples/sls/store.yaml:5]:
```suggestion
```
Resource is cluster-scoped :+1:

File: apis/sls/v1alpha1/store_types.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	xpv1.ResourceStatus `json:",inline"`
+	AtProvider          StoreObservation `json:"atProvider,omitempty"`
+}
+
+// StoreParameters define the desired state of an SLS store.
+type StoreParameters struct {
+	// SLS project name
+	// +kubebuilder:validation:MinLength:=3
+	// +kubebuilder:validation:MaxLength:=63
+	ProjectName string `json:"projectName"`
```
Comment (hasheddan) [inline apis/sls/v1alpha1/store_types.go:54]: I believe you will want cross resource referencing implemented for this field to that folks are able to point to their `Project` resource in the cluster to populate the `ProjectName`. Let me know if you have any questions around implementation :)

File: pkg/controller/sls/store_controller.go (hunk 2 above)
Comment (hasheddan) [reply pkg/controller/sls/store_controller.go:81]: I see this has already been implemented in https://github.com/crossplane/provider-alibaba/pull/64, apologies for the delayed review there! I believe it needs a rebase but I will prioritize getting eyes on that!

File: apis/sls/v1alpha1/store_types.go (hunk 5 above)
Comment (zzxwill) [reply apis/sls/v1alpha1/store_types.go:54]: Correct. That's exactly what I need. I noticed a design doc on cross resource referencing https://github.com/crossplane/crossplane/blob/master/design/one-pager-cross-resource-referencing.md. A more straight example would be much appreciated.

Comment (zzxwill) [reply apis/sls/v1alpha1/store_types.go:54]: @hasheddan I thought it over. Resource SLS Store can also be independent of SLS Project. If the project doesn't exist, a `ProjectNotExist` error can occur https://www.alibabacloud.com/help/zh/doc-detail/29015.htm which is acceptable.

**Review by wonderflow — APPROVED** [review]
@hasheddan Thanks dan, it's worked now, and let's refine in following PR. @zzxwill  is also alibaba cloud staff and our community maintainer.

**Pull request merged**